### DISCUSSION TITLE
Визуализация очередности ходов и структуры игры

### DISCUSSION THREAD

**Discussion by nergal-perm** [nergal-perm/tg-core#34] [Вопросы]
Каким способом можно наглядно представить структуру игры? Очередность ходов, возможность выбора одного из нескольких действий, заданные последовательности действий, вот это вот все?

Можно попробовать накидывать варианты визуализации конкретных игр в Miro.

А для описания структуры, мне кажется, вполне подходит XML. Начал набрасывать [черновик такой нотации в wiki](https://github.com/nergal-perm/tg-core/wiki/%D0%9C%D0%BE%D0%B4%D0%B5%D0%BB%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5-%D1%85%D0%BE%D0%B4%D0%BE%D0%B2-%D0%B8-%D1%80%D0%B0%D1%83%D0%BD%D0%B4%D0%BE%D0%B2).

**Comment by nergal-perm** [comment]
Пришла мысль, что структура игры - это больше о том, КОГДА дать игроку возможность совершить действие. А вот ЧТО именно может сделать игрок - за это отвечают правила игры. Это разделение ответственности.

Как это влияет на визуализацию?

Во-первых, с точки зрения структуры игры не важны конкретные типы действий. В структуре должно отразиться просто действие, без конкретики относительно его характера. Зато важно указать, сколько раз выполняется действие, каков идентификатор текущего хода или раунда, до или после действия выполняются проверки.

Во-вторых, в структуре не нужны элементы типа "выбор", так как списки доступных действий и необходимость сделать выбор - это знания и ответственность правил.

**Reply by citerm** [reply]
Мне кажется в свете нашего обсуждения 2.04.20 получается следующая картина:

- список доступных действий определяется способностями
- способности определяются состоянием игры в целом
- количество выполнений действий и другие "казалось бы" циклические правила выполнения действий, тоже могут быть урегулированы способностями, пример: ход игрока состоит из 2х действий, после первого действия состояние игры изменилось и счетчик доступных в этом ходу действий у игрока уменьшился, способности все еще позволяют ему совершить действие. После второго действия счетчик доступных действий исчерпал себя, после расчета способностей игрок больше не может совершать действий, ход "передан" другому игроку, так как способности разрешают соверать действия уже другому игроку.
